lib/include/pl/core/ast/ast_node_type_decl.hpp

```diff
@@ -76,15 +76,46 @@ namespace pl::core::ast {
         [[nodiscard]] std::vector<std::shared_ptr<ptrn::Pattern>> createPatterns(Evaluator *evaluator) const override {
             evaluator->updateRuntime(this);
 
-            evaluator->pushTemplateParameters();
-            ON_SCOPE_EXIT { evaluator->popTemplateParameters(); };
+            std::vector<std::unique_ptr<ASTNodeLiteral>> templateParamLiterals(this->m_templateParameters.size());
 
-            for (const auto &templateParameter : this->m_templateParameters) {
+            // Get template parameter values before pushing a new template parameter scope so that variables from the parent scope are used before being overwritten.
+            for (size_t i = 0; i < this->m_templateParameters.size(); i++) {
```

Comment:
This pattern here can generally be replaced with a for-each loop. `for (const auto &templateParameter : this->m_templateParameters)`

Reply:
This uses an index so that it can assign to the vector at the correct parameter index, leaving null pointers in all the other slots. I could change it to use for-each and an index outside the scope of the loop, though, up to you.